File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.cc
// segment tree
#include <bits/stdc++.h>

using namespace std;

int N, M, K;
vector<long long> tree;
vector<long long> arr;

long long init(int node, int start, int end) {
	// 리프 노드인 경우
	if (start == end) {
		return tree[node] = arr[start];
	} else {
		// node의 left child = node*2
		// node의 right child = node*2+1
		// node의 구간이 [start,end]이면
		// left child 구간 = [start,(start+end)/2]
		// right child 구간 = [(start+end)/2+1,end]
		return tree[node] = init(node * 2, start, (start + end) / 2) +
							init(node * 2 + 1, (start + end) / 2 + 1, end);
	}
}

long long sum(int node, int start, int end, int left, int right) {
	// [left,right]와 [start,end]가 겹치지 않는 경우
	if (left > end || right < start) {
		return 0;
	}
	// [left,right]가 [start,end]를 완전히 포함하는 경우
	if (left <= start && end <= right) {
		return tree[node];
	}
	return sum(node * 2, start, (start + end) / 2, left, right) +
		   sum(node * 2 + 1, (start + end) / 2 + 1, end, left, right);
}

void update(int node, int start, int end, int index, long long diff) {
	// [start,end]에 index가 포함되지 않는 경우
	if (index < start || index > end) return;

	tree[node] = tree[node] + diff;
	if (start != end) {
		update(node * 2, start, (start + end) / 2, index, diff);
		update(node * 2 + 1, (start + end) / 2 + 1, end, index, diff);
	}
}

void update(int index, long long val) {
	long long diff = val - arr[index];
	arr[index] = val;
	update(1, 0, N - 1, index, diff);
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);

	cin >> N >> M >> K;

	tree.resize((1 << ((int) ceil(log2(N)) + 1)));
	arr.resize(N);

	for (int i = 0; i < N; i++) {
		cin >> arr[i];
	}

	init(1, 0, N - 1);

	int a, b;
	for (int i = 0; i < M + K; i++) {
		cin >> a >> b;
		if (a == 1) {
			long long c;
			cin >> c;
			update(b - 1, c);
		} else if (a == 2) {
			int c;
			cin >> c;
			printf("%lld\n", sum(1, 0, N - 1, b - 1, c - 1));
		}
	}

	return 0;
}

File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.js
const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});
const input = [];

rl.on('line', (line) => {
  input.push(line);
}).on('close', () => {
  const [N, M, K] = input[0].split(' ').map(Number);
  const numbers = input.slice(1, N + 1).map(BigInt);
  const operations = input.slice(N + 1, N + M + K + 1).map((v) => v.split(' ').map(Number));

  solution(numbers, operations);
});

const solution = (numbers, operations) => {
  const segTree = new SegmentTree(numbers);
  const N = numbers.length;
  const result = [];

  segTree.init(1, 0, N - 1);

  for (const [a, b, c] of operations) {
    if (a === 1) {
      segTree.update(b - 1, BigInt(c));
    } else {
      result.push(segTree.sum(1, 0, N - 1, b - 1, c - 1));
    }
  }

  console.log(result.join('\n'));
};

class SegmentTree {
  constructor(numbers) {
    this.tree = Array(1 << Math.ceil(Math.log2(numbers.length) + 1)).fill(0n);
    this.numbers = numbers;
  }

  init(node, start, end) {
    if (start === end) return (this.tree[node] = this.numbers[start]);

    const mid = Math.floor((start + end) / 2);
    return (this.tree[node] =
      this.init(node * 2, start, mid) + this.init(node * 2 + 1, mid + 1, end));
  }

  sum(node, start, end, left, right) {
    if (left > end || right < start) return 0n;
    if (left <= start && end <= right) return this.tree[node];

    const mid = Math.floor((start + end) / 2);
    return (
      this.sum(node * 2, start, mid, left, right) +
      this.sum(node * 2 + 1, mid + 1, end, left, right)
    );
  }

  updateRecursive(node, start, end, index, diff) {
    if (index < start || index > end) return;

    this.tree[node] += diff;

    if (start === end) return;

    const mid = Math.floor((start + end) / 2);
    this.updateRecursive(node * 2, start, mid, index, diff);
    this.updateRecursive(node * 2 + 1, mid + 1, end, index, diff);
  }

  update(index, value) {
    const diff = value - this.numbers[index];
    this.numbers[index] = value;
    this.updateRecursive(1, 0, this.numbers.length - 1, index, diff);
  }
}
